src/ssl.c

```diff
@@ -73,41 +81,165 @@ ssl_startup(void)
 	pthread_mutex_unlock(&ssl_lock);
 }
 
+static validate_result
+matches_common_name(const char *hostname, const X509 *server_cert)
+{
+	int common_name_loc = -1;
+	X509_NAME_ENTRY *common_name_entry = NULL;
+	ASN1_STRING *common_name_asn1 = NULL;
+	char *common_name_str = NULL;
+
+	// Find the position of the CN field in the Subject field of the certificate
+	common_name_loc = X509_NAME_get_index_by_NID(X509_get_subject_name((X509 *) server_cert), NID_commonName, -1);
+	if (common_name_loc < 0) {
+		return Error;
+	}
+
+	// Extract the CN field
+	common_name_entry = X509_NAME_get_entry(X509_get_subject_name((X509 *) server_cert), common_name_loc);
+	if (common_name_entry == NULL) {
+		return Error;
+	}
+
+	// Convert the CN field to a C string
+	common_name_asn1 = X509_NAME_ENTRY_get_data(common_name_entry);
+	if (common_name_asn1 == NULL) {
+		return Error;
+	}
+	#if OPENSSL_VERSION_NUMBER < 0x10101000L
+		common_name_str = (char *) ASN1_STRING_data(common_name_asn1);
+	#else
+		common_name_str = (char *) ASN1_STRING_get0_data(common_name_asn1);
+	#endif
```

Comment:
These functions need to be aliased somewhere else.

Reply:
I have `#define`ed a `ASN1_STRING_data_compat` that abstracts the API differences.